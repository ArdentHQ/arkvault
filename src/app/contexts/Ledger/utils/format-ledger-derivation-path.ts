import { LedgerDerivationScheme } from "@/app/contexts/Ledger/Ledger.contracts";

export const formatLedgerDerivationPath = (scheme: LedgerDerivationScheme) =>
	`m/${scheme.purpose || 44}'/${scheme.coinType}'/${scheme.account || 0}'/${scheme.change || 0}/${
		scheme.address || 0
	}`;
