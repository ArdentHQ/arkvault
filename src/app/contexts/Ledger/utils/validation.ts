import semver from "semver";
import { Coins } from "@payvo/sdk";
import { minVersionList } from "@/app/contexts/Ledger/Ledger.contracts";

export const hasRequiredAppVersion = async (coin: Coins.Coin) => {
	const coinId = coin.network().coin();

	if (minVersionList[coinId]) {
		const currentVersion = await coin.ledger().getVersion();

		if (semver.lt(currentVersion, minVersionList[coinId])) {
			return false;
		}
	}

	return true;
};
