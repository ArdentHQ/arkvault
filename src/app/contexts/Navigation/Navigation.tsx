import React, { useState, createContext, useContext, useMemo } from "react";

interface NavigationContextType {
	setHasFixedFormButtons: (show: boolean) => void;
	setShowMobileNavigation: (show: boolean) => void;
	hasFixedFormButtons: boolean;
	showMobileNavigation: boolean;
}

interface Properties {
	children: React.ReactNode;
}

const NavigationContext = createContext<NavigationContextType | undefined>(undefined);

export const NavigationProvider = ({ children }: Properties) => {
	// Places like the wallet creation form contains fixed buttons floating at
	// the bottom, we use this state to change the design of the mobile menu
	// to prevent the buttons from overlapping with the menu.
	const [hasFixedFormButtons, setHasFixedFormButtons] = useState(false);
	const [showMobileNavigation, setShowMobileNavigation] = useState(true);

	const navigationContext = useMemo<NavigationContextType>(
		() => ({
			hasFixedFormButtons,
			setHasFixedFormButtons,
			setShowMobileNavigation,
			showMobileNavigation,
		}),
		[hasFixedFormButtons, showMobileNavigation],
	);

	return <NavigationContext.Provider value={navigationContext}>{children}</NavigationContext.Provider>;
};

export const useNavigationContext = () => {
	const value = useContext(NavigationContext);

	if (value === undefined) {
		throw new Error("[useNavigationContext] Component not wrapped within a Provider");
	}

	return value;
};
