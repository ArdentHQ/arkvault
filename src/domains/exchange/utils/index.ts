export * from "./assertions";
export * from "./errors";
