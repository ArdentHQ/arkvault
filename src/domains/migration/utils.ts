import { MigrationTransactionStatus } from "@/domains/migration/migration.contracts";

export const getIcon = (status: MigrationTransactionStatus) => {
	if (status === MigrationTransactionStatus.Confirmed) {
		return {
			color: "text-theme-success-600",
			name: "CircleCheckMark",
		};
	}

	return {
		color: "text-theme-warning-300",
		name: "Clock",
	};
};
