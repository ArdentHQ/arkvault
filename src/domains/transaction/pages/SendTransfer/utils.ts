export const getFeeType = (numberOfRecipients: number): "multiPayment" | "transfer" =>
	numberOfRecipients > 1 ? "multiPayment" : "transfer";
