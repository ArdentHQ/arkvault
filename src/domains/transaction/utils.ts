import { Services } from "@ardenthq/sdk";
import { RecipientItem } from "@/domains/transaction/components/RecipientList/RecipientList.contracts";

export const isNoDeviceError = (error: any) => {
	if (!error) {
		return false;
	}

	if (String(error).includes("no device found")) {
		return true;
	}

	return false;
};

export const isRejectionError = (error: any) => String(error).includes("Condition of use not satisfied");

export const handleBroadcastError = ({ errors }: Services.BroadcastResponse) => {
	const allErrors = Object.values(errors);

	if (allErrors.length === 0) {
		return;
	}

	throw new Error(allErrors[0]);
};

export const getTransferType = ({ recipients }: { recipients: RecipientItem[] }): "multiPayment" | "transfer" => {
	const isMultiPayment = recipients.length > 1;
	return isMultiPayment ? "multiPayment" : "transfer";
};

export const withAbortPromise =
	(signal?: AbortSignal, callback?: () => void) =>
	<T>(promise: Promise<T>) =>
		new Promise<T>((resolve, reject) => {
			if (signal) {
				signal.addEventListener("abort", () => {
					callback?.();
					reject("ERR_ABORT");
				});
			}

			return promise.then(resolve).catch(reject);
		});
