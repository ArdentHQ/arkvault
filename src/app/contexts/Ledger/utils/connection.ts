import retry from "async-retry";
import { Coins } from "@ardenthq/sdk";
import { formatLedgerDerivationPath } from "./format-ledger-derivation-path";
import { hasRequiredAppVersion } from "./validation";

export const accessLedgerDevice = async (coin: Coins.Coin) => {
	try {
		await coin.__construct();
		await coin.ledger().connect();
	} catch (error) {
		// If the device is open, continue normally.
		// Can be triggered when the user retries ledger connection.
		if (error.message !== "The device is already open.") {
			throw error;
		}
	}
};

export const accessLedgerApp = async ({ coin }: { coin: Coins.Coin }) => {
	await accessLedgerDevice(coin);

	if (!(await hasRequiredAppVersion(coin))) {
		throw new Error("VERSION_ERROR");
	}

	// Ensure that the app is accessible.
	await coin.ledger().getPublicKey(
		formatLedgerDerivationPath({
			coinType: coin.config().get<number>("network.constants.slip44"),
		}),
	);
};

export const persistLedgerConnection = async ({
	coin,
	options,
	hasRequestedAbort,
}: {
	coin: Coins.Coin;
	options: retry.Options;
	hasRequestedAbort: () => boolean;
}) => {
	const retryAccess: retry.RetryFunction<void> = async (bail: (error: Error) => void, attempts: number) => {
		if (hasRequestedAbort() && attempts > 1) {
			bail(new Error("CONNECTION_ERROR"));
			return;
		}

		try {
			await accessLedgerApp({ coin });
		} catch (error) {
			// Bail on version error or continue retrying access.
			if (error.message === "VERSION_ERROR") {
				bail(new Error("VERSION_ERROR"));
			}

			throw error;
		}
	};

	await retry(retryAccess, options);
};
