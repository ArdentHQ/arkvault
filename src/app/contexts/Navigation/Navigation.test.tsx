import userEvent from "@testing-library/user-event";
import React from "react";

import { NavigationProvider, useNavigationContext } from "./Navigation";
import { render, screen, waitFor } from "@/utils/testing-library";

describe("Navigation Context", () => {
	it("should render the wrapper properly", () => {
		const { container, asFragment } = render(
			<NavigationProvider>
				<span data-testid="NavigationProvider__content">Navigation Provider content</span>
			</NavigationProvider>,
		);

		expect(screen.getByTestId("NavigationProvider__content")).toBeInTheDocument();

		expect(container).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	it("should throw without provider", () => {
		const consoleSpy = jest.spyOn(console, "error").mockImplementation(() => {});

		const Test = () => {
			useNavigationContext();
			return <p>Navigation content</p>;
		};

		expect(() => render(<Test />, { withProviders: false })).toThrow(
			"[useNavigationContext] Component not wrapped within a Provider",
		);

		consoleSpy.mockRestore();
	});

	it("should update navigation", async () => {
		const Test = () => {
			const { hasFixedFormButtons, setHasFixedFormButtons } = useNavigationContext();
			return (
				<div data-testid="Navigation__consumer" onClick={() => setHasFixedFormButtons(true)}>
					{hasFixedFormButtons ? "Has fixed form buttons" : "Does not have fixed form buttons"}
				</div>
			);
		};

		const { asFragment } = render(
			<NavigationProvider>
				<Test />
			</NavigationProvider>,
		);

		expect(screen.getByTestId("Navigation__consumer")).toBeInTheDocument();

		expect(screen.getByText("Does not have fixed form buttons")).toBeInTheDocument();

		userEvent.click(screen.getByTestId("Navigation__consumer"));

		await waitFor(() => expect(screen.queryByText("Does not have fixed form buttons")).not.toBeInTheDocument());

		expect(asFragment()).toMatchSnapshot();
	});
});
