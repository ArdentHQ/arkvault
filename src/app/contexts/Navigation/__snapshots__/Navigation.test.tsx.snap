// Vitest Snapshot v1

exports[`Navigation Context > should render the wrapper properly 1`] = `
<DocumentFragment>
  <span
    data-testid="NavigationProvider__content"
  >
    Navigation Provider content
  </span>
</DocumentFragment>
`;

exports[`Navigation Context > should update navigation 1`] = `
<DocumentFragment>
  <div
    data-testid="Navigation__consumer"
  >
    Has fixed form buttons
  </div>
</DocumentFragment>
`;
